harvester/harvester
# feat: add webhook for snapshot space management

**IMPORTANT: Please do not create a Pull Request without creating an issue first.**

**Problem:**
Snapshots use extra storage. Currently, users don't have a way to control snapshots count/size and check how much size snapshot uses. Snapshots may use more storage than users expect.

**Solution:**
~Introduce `harvesterhcio.io/totalSnapshotSize` annotation to VM and Namespace. Users can set a value on it to limit total snapshot usage in a VM or Namespace.~

Introduce `ResourceQuota` to control total namespace snapshot size and VM snapshot size.

**Related Issue:**
https://github.com/harvester/harvester/issues/4478, https://github.com/harvester/harvester/pull/4497

**Test plan:**

1. `namespaceTotalSnapshotSizeUsage` and `vmTotalSnapshotSizeUsage` values format.
    - Deny value which can't be parse by [ParseQuantity](https://github.com/kubernetes/apimachinery/blob/e126c655b8146add9b9c19bf89dab6eb85578814/pkg/api/resource/quantity.go#L276-L277).
    - Deny negative value.
2. Set `30Gi` limit on a VM after it's created. The VM has two volumes. The first volume size is `10Gi` and the second volume size is `20Gi`.
    - Write data to make two volumes full. Create the first VMBackup. It should be allowed. The total snapshot size is `30Gi`.
    - Clear data and write data to make two volumes full again. Create the seoncd VMBackup. It should be allowed, because we haven't set the annotation. The total snapshot size is `60Gi`.
    - Set `30Gi` on the VM. It should be allowed.
    - Clear data and write data to make two volumes full again.  Create the third VMBackup. It should be denied, because the total snapshot size will be `90Gi` if the request is allowed.
    - Remove the first and the second VMBackup. The total snapshot size is `0Gi`.
    - Create a new VMBackup. It should be allowed. The total snapshot size is `30Gi`.
4. Set `10Gi` limit on a Namespace. Create a VM with size `10Gi`.
    - Write data to make the volume full. Create the first VMBackup. It should be allowed. The total snapshot size is `10Gi`.
    - Clear data and write data to make the volume full again. Create the second VMBackup. It should be denied, because the total snapshot size will be `20Gi` if the request is allowed.
    - Remove the first VMBackup. The total snapshot size is `0Gi`.
    - Create a new VMBackup. It should be allowed. The total snapshot size is `10Gi`.
5. Set `10Gi` limit on a Namespace. Create a volume with size `10Gi`.
    - Write data to make the volume full. Create the first VolumeSnapshot. It should be allowed. The total snapshot size is `10Gi`.
    - Clear data and write data to make the volume full again. Create the second VolumeSnapshot. It should be denied, because the total snapshot size will be `20Gi` if the request is allowed.
    - Remove the first VolumeSnapshot. The total snapshot size is `0Gi`.
    - Create a new VolumeSnapshot. It should be allowed. The total snapshot size is `10Gi`.
6. Users with `Harvester ResourceQuotas Read Only` role.
    - Apply `harvester-resourcequotas-admin` and `harvester-resourcequotas-read-only` to a Rancher.
    - Import Harvester to a Rancher.
    - Create a new user with `Harvester ResourceQuotas Read Only` role.
    - Add the user to a project as member.
    - Send requests to update Resource Quotas via API on Namespace / VM. Both requests should be denied.
7. Users with `Harvester ResourceQuotas Admin` role.
    - Apply `harvester-resourcequotas-admin` and `harvester-resourcequotas-read-only` to a Rancher.
    - Import Harvester to a Rancher.
    - Create a new user with `Harvester ResourceQuotas Admin` role.
    - Add the user to a project as readonly user.
    - Send requests to update Resource Quotas via API on Namespace / VM. Both requests should be allowed.
